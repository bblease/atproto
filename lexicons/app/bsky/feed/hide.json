{
  "lexicon": 1,
  "id": "app.bsky.feed.hide",
  "defs": {
    "main": {
      "type": "procedure",
      "description": "Hide a post from future timelines.",
      "input": {
        "encoding": "application/json",
        "schema": {
          "type": "object",
          "required": ["uri"],
          "properties": {
            "uri": { "type": "string" }
          }
        }
      }
    }
  }
}
