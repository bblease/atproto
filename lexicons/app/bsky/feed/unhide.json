{
  "lexicon": 1,
  "id": "app.bsky.feed.unhide",
  "defs": {
    "main": {
      "type": "procedure",
      "description": "Unhide a post from future timelines.",
      "input": {
        "encoding": "application/json",
        "schema": {
          "type": "object",
          "required": ["uri"],
          "properties": {
            "uri": { "type": "string" }
          }
        }
      }
    }
  }
}
