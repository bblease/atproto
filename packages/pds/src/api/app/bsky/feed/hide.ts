import { Server } from '../../../../lexicon'
import { InvalidRequestError } from '@atproto/xrpc-server'
import AppContext from '../../../../context'

export default function (server: Server, ctx: AppContext) {
  server.app.bsky.feed.hide({
    auth: ctx.accessVerifier,
    handler: async ({ auth, input }) => {
      const { uri } = input.body
      const requester = auth.credentials.did
      const { db, services } = ctx
      
      await services.feed(db).hide({
        uri,
        hiddenByDid: requester,
      })
    },
  })
}

