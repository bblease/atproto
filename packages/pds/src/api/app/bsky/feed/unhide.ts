import { Server } from '../../../../lexicon'
import { InvalidRequestError } from '@atproto/xrpc-server'
import AppContext from '../../../../context'

export default function (server: Server, ctx: AppContext) {
  return
}
