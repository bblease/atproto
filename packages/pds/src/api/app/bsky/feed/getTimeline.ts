import { InvalidRequestError } from '@atproto/xrpc-server'
import { Server } from '../../../../lexicon'
import { FeedAlgorithm, FeedKeyset, composeFeed } from '../util/feed'
import { paginate } from '../../../../db/pagination'
import AppContext from '../../../../context'
import { FeedRow } from '../../../../services/feed'

export default function (server: Server, ctx: AppContext) {
  server.app.bsky.feed.getTimeline({
    auth: ctx.accessVerifier,
    handler: async ({ params, auth }) => {
      const { algorithm, limit, before } = params
      const db = ctx.db.db
      const { ref } = db.dynamic
      const requester = auth.credentials.did

      const feedService = ctx.services.feed(ctx.db)

      // which users are muted?
      const mutedQb = db
        .selectFrom('mute')
        .select('did')
        .where('mutedByDid', '=', requester)

      const hiddenQb = db
        .selectFrom('hide')
        .select('uri')
        .where('hiddenByDid', '=', requester)

      let postQb
      let repostQb
      switch (algorithm) {
        case FeedAlgorithm.ReverseChronological:
          const followingIdsSubquery = db
            .selectFrom('follow')
            .select('follow.subjectDid')
            .where('follow.creator', '=', requester)

          postQb = feedService
            .selectPostQb()
            .where((qb) =>
              qb
                .where('creator', '=', requester)
                .orWhere('creator', 'in', followingIdsSubquery),
            )
            .whereNotExists(mutedQb.whereRef('did', '=', ref('post.creator'))) // Hide posts of muted actors
            .whereNotExists(hiddenQb.whereRef('uri', '=', ref('post.uri'))) // Remove hidden posts

          repostQb = feedService
            .selectRepostQb()
            .where((qb) =>
              qb
                .where('repost.creator', '=', requester)
                .orWhere('repost.creator', 'in', followingIdsSubquery),
            )
            .whereNotExists(
              mutedQb
                .whereRef('did', '=', ref('post.creator')) // Hide reposts of or by muted actors
                .orWhereRef('did', '=', ref('repost.creator')),
            )
            .whereNotExists(
              hiddenQb
                .whereRef('uri', '=', ref('post.uri'))
                .orWhereRef('uri', '=', ref('repost.uri')),
            )
          break

        case FeedAlgorithm.AllPosts:
          postQb = feedService
            .selectPostQb()
            .whereNotExists(mutedQb.whereRef('did', '=', ref('post.creator'))) // Hide posts of muted actors
            .whereNotExists(hiddenQb.whereRef('uri', '=', ref('post.uri'))) // Remove hidden posts

          repostQb = feedService
            .selectRepostQb()
            .whereNotExists(
              mutedQb
                .whereRef('did', '=', ref('post.creator')) // Hide reposts of or by muted actors
                .orWhereRef('did', '=', ref('repost.creator')),
            )
            .whereNotExists(
              hiddenQb
                .whereRef('uri', '=', ref('post.uri'))
                .orWhereRef('uri', '=', ref('repost.uri')),
            )
          break
        default:
          throw new InvalidRequestError(`Unsupported algorithm: ${algorithm}`)
      }

      const keyset = new FeedKeyset(ref('cursor'), ref('postCid'))
      let feedItemsQb = db
        .selectFrom(postQb.union(repostQb).as('feed_items'))
        .selectAll()
      feedItemsQb = paginate(feedItemsQb, {
        limit,
        before,
        keyset,
      })

      const feedItems: FeedRow[] = (await feedItemsQb.execute()) as FeedRow[]
      const feed = await composeFeed(feedService, feedItems, requester)

      return {
        encoding: 'application/json',
        body: {
          feed,
          cursor: keyset.packFromResult(feedItems),
        },
      }
    },
  })
}
