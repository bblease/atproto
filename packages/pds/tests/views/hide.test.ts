import AtpAgent from '@atproto/api'
import { Main as FeedViewPost } from '../../src/lexicon/types/app/bsky/feed/feedViewPost'
import {
  runTestServer,
  forSnapshot,
  CloseFn,
  getOriginator,
  adminAuth,
} from '../_util'
import { SeedClient } from '../seeds/client'
import basicSeed from '../seeds/basic'
import { FeedAlgorithm } from '../../src/api/app/bsky/util/feed'

describe('hide certain posts from the user feed', () => {
  let agent: AtpAgent
  let close: CloseFn
  let sc: SeedClient

  // account dids, for convenience
  let alice: string
  let bob: string
  let dan: string

  beforeAll(async () => {
    const server = await runTestServer({
      dbPostgresSchema: 'views_home_feed',
    })
    close = server.close
    agent = new AtpAgent({ service: server.url })
    sc = new SeedClient(agent)
    await basicSeed(sc, server.ctx.messageQueue)
    alice = sc.dids.alice
    bob = sc.dids.bob
    dan = sc.dids.dan
  })

  afterAll(async () => {
    await close()
  })

  it('hides a post from the user feed.', async () => {
    const postUri1 = sc.posts[dan][1].ref.uriStr // Repost
    const postUri2 = sc.replies[bob][0].ref.uriStr // Post and reply parent
    
    const actionResults = await Promise.all(
      [postUri1, postUri2].map((postUri) =>
        agent.api.app.bsky.feed.hide(
          {
            uri: postUri
          },
          {
            encoding: 'application/json',
            headers: sc.getHeaders(alice),
          },
        ),
      ),
    )

    const aliceTL = await agent.api.app.bsky.feed.getTimeline(
      { algorithm: FeedAlgorithm.ReverseChronological },
      { headers: sc.getHeaders(alice) },
    )

    expect(forSnapshot(aliceTL.data.feed)).toMatchSnapshot()
  })
})
