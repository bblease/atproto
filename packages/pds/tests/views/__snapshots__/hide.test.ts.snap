// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hide certain posts from the user feed hides a post from the user feed. 1`] = `
[
  {
    "post": {
      "author": {
        "avatar": "https://pds.public.url/image/KzkHFsMRQ6oAKCHCRKFA1H-rDdc7VOtvEVpUJ82TwyQ/rs:fill:1000:1000:1:0/plain/bafkreiaivizp4xldojmmpuzmiu75cmea7nq56dnntnuhzhsjcb63aou5ei@jpeg",
        "declaration": {
          "actorType": "app.bsky.system.actorUser",
          "cid": "cids(1)",
        },
        "did": "user(0)",
        "displayName": "ali",
        "handle": "alice.test",
        "viewer": {
          "muted": false,
        },
      },
      "cid": "cids(0)",
      "downvoteCount": 1,
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "record": {
        "$type": "app.bsky.feed.post",
        "createdAt": "1970-01-01T00:00:00.000Z",
        "text": "again",
      },
      "replyCount": 2,
      "repostCount": 1,
      "upvoteCount": 2,
      "uri": "record(0)",
      "viewer": {},
    },
    "reason": {
      "$type": "app.bsky.feed.feedViewPost#reasonRepost",
      "by": {
        "declaration": {
          "actorType": "app.bsky.system.actorUser",
          "cid": "cids(1)",
        },
        "did": "user(1)",
        "handle": "dan.test",
        "viewer": {
          "following": "record(1)",
          "muted": false,
        },
      },
      "indexedAt": "1970-01-01T00:00:00.000Z",
    },
  },
  {
    "post": {
      "author": {
        "declaration": {
          "actorType": "app.bsky.system.actorUser",
          "cid": "cids(1)",
        },
        "did": "user(1)",
        "handle": "dan.test",
        "viewer": {
          "following": "record(1)",
          "muted": false,
        },
      },
      "cid": "cids(2)",
      "downvoteCount": 0,
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "record": {
        "$type": "app.bsky.feed.post",
        "createdAt": "1970-01-01T00:00:00.000Z",
        "entities": [
          {
            "index": {
              "end": 18,
              "start": 0,
            },
            "type": "mention",
            "value": "user(0)",
          },
        ],
        "text": "@alice.bluesky.xyz is the best",
      },
      "replyCount": 0,
      "repostCount": 1,
      "upvoteCount": 0,
      "uri": "record(2)",
      "viewer": {},
    },
    "reason": {
      "$type": "app.bsky.feed.feedViewPost#reasonRepost",
      "by": {
        "declaration": {
          "actorType": "app.bsky.system.actorUser",
          "cid": "cids(1)",
        },
        "did": "user(2)",
        "handle": "carol.test",
        "viewer": {
          "followedBy": "record(4)",
          "following": "record(3)",
          "muted": false,
        },
      },
      "indexedAt": "1970-01-01T00:00:00.000Z",
    },
  },
  {
    "post": {
      "author": {
        "avatar": "https://pds.public.url/image/KzkHFsMRQ6oAKCHCRKFA1H-rDdc7VOtvEVpUJ82TwyQ/rs:fill:1000:1000:1:0/plain/bafkreiaivizp4xldojmmpuzmiu75cmea7nq56dnntnuhzhsjcb63aou5ei@jpeg",
        "declaration": {
          "actorType": "app.bsky.system.actorUser",
          "cid": "cids(1)",
        },
        "did": "user(0)",
        "displayName": "ali",
        "handle": "alice.test",
        "viewer": {
          "muted": false,
        },
      },
      "cid": "cids(3)",
      "downvoteCount": 0,
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "record": {
        "$type": "app.bsky.feed.post",
        "createdAt": "1970-01-01T00:00:00.000Z",
        "reply": {
          "parent": {
            "cid": "cids(4)",
            "uri": "record(6)",
          },
          "root": {
            "cid": "cids(0)",
            "uri": "record(0)",
          },
        },
        "text": "thanks bob",
      },
      "replyCount": 0,
      "repostCount": 0,
      "upvoteCount": 0,
      "uri": "record(5)",
      "viewer": {},
    },
    "reply": {
      "parent": {
        "author": {
          "avatar": "https://pds.public.url/image/KzkHFsMRQ6oAKCHCRKFA1H-rDdc7VOtvEVpUJ82TwyQ/rs:fill:1000:1000:1:0/plain/bafkreiaivizp4xldojmmpuzmiu75cmea7nq56dnntnuhzhsjcb63aou5ei@jpeg",
          "declaration": {
            "actorType": "app.bsky.system.actorUser",
            "cid": "cids(1)",
          },
          "did": "user(3)",
          "displayName": "bobby",
          "handle": "bob.test",
          "viewer": {
            "followedBy": "record(8)",
            "following": "record(7)",
            "muted": false,
          },
        },
        "cid": "cids(4)",
        "downvoteCount": 0,
        "embed": {
          "$type": "app.bsky.embed.images#presented",
          "images": [
            {
              "alt": "tests/image/fixtures/key-landscape-small.jpg",
              "fullsize": "https://pds.public.url/image/AiDXkxVbgBksxb1nfiRn1m6S4K8_mee6o8r-UGLNzOM/rs:fit:2000:2000:1:0/plain/bafkreigy5p3xxceipk2o6nqtnugpft26ol6yleqhboqziino7axvdngtci@jpeg",
              "thumb": "https://pds.public.url/image/uc7FGfiGv0mMqmk9XiqHXrIhNymLHaex7Ge8nEhmXqo/rs:fit:1000:1000:1:0/plain/bafkreigy5p3xxceipk2o6nqtnugpft26ol6yleqhboqziino7axvdngtci@jpeg",
            },
          ],
        },
        "indexedAt": "1970-01-01T00:00:00.000Z",
        "record": {
          "$type": "app.bsky.feed.post",
          "createdAt": "1970-01-01T00:00:00.000Z",
          "embed": {
            "$type": "app.bsky.embed.images",
            "images": [
              {
                "alt": "tests/image/fixtures/key-landscape-small.jpg",
                "image": {
                  "cid": "cids(5)",
                  "mimeType": "image/jpeg",
                },
              },
            ],
          },
          "reply": {
            "parent": {
              "cid": "cids(0)",
              "uri": "record(0)",
            },
            "root": {
              "cid": "cids(0)",
              "uri": "record(0)",
            },
          },
          "text": "hear that",
        },
        "replyCount": 1,
        "repostCount": 0,
        "upvoteCount": 0,
        "uri": "record(6)",
        "viewer": {},
      },
      "root": {
        "author": {
          "avatar": "https://pds.public.url/image/KzkHFsMRQ6oAKCHCRKFA1H-rDdc7VOtvEVpUJ82TwyQ/rs:fill:1000:1000:1:0/plain/bafkreiaivizp4xldojmmpuzmiu75cmea7nq56dnntnuhzhsjcb63aou5ei@jpeg",
          "declaration": {
            "actorType": "app.bsky.system.actorUser",
            "cid": "cids(1)",
          },
          "did": "user(0)",
          "displayName": "ali",
          "handle": "alice.test",
          "viewer": {
            "muted": false,
          },
        },
        "cid": "cids(0)",
        "downvoteCount": 1,
        "indexedAt": "1970-01-01T00:00:00.000Z",
        "record": {
          "$type": "app.bsky.feed.post",
          "createdAt": "1970-01-01T00:00:00.000Z",
          "text": "again",
        },
        "replyCount": 2,
        "repostCount": 1,
        "upvoteCount": 2,
        "uri": "record(0)",
        "viewer": {},
      },
    },
  },
  {
    "post": {
      "author": {
        "declaration": {
          "actorType": "app.bsky.system.actorUser",
          "cid": "cids(1)",
        },
        "did": "user(2)",
        "handle": "carol.test",
        "viewer": {
          "followedBy": "record(4)",
          "following": "record(3)",
          "muted": false,
        },
      },
      "cid": "cids(6)",
      "downvoteCount": 0,
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "record": {
        "$type": "app.bsky.feed.post",
        "createdAt": "1970-01-01T00:00:00.000Z",
        "reply": {
          "parent": {
            "cid": "cids(0)",
            "uri": "record(0)",
          },
          "root": {
            "cid": "cids(0)",
            "uri": "record(0)",
          },
        },
        "text": "of course",
      },
      "replyCount": 0,
      "repostCount": 0,
      "upvoteCount": 0,
      "uri": "record(9)",
      "viewer": {},
    },
    "reply": {
      "parent": {
        "author": {
          "avatar": "https://pds.public.url/image/KzkHFsMRQ6oAKCHCRKFA1H-rDdc7VOtvEVpUJ82TwyQ/rs:fill:1000:1000:1:0/plain/bafkreiaivizp4xldojmmpuzmiu75cmea7nq56dnntnuhzhsjcb63aou5ei@jpeg",
          "declaration": {
            "actorType": "app.bsky.system.actorUser",
            "cid": "cids(1)",
          },
          "did": "user(0)",
          "displayName": "ali",
          "handle": "alice.test",
          "viewer": {
            "muted": false,
          },
        },
        "cid": "cids(0)",
        "downvoteCount": 1,
        "indexedAt": "1970-01-01T00:00:00.000Z",
        "record": {
          "$type": "app.bsky.feed.post",
          "createdAt": "1970-01-01T00:00:00.000Z",
          "text": "again",
        },
        "replyCount": 2,
        "repostCount": 1,
        "upvoteCount": 2,
        "uri": "record(0)",
        "viewer": {},
      },
      "root": {
        "author": {
          "avatar": "https://pds.public.url/image/KzkHFsMRQ6oAKCHCRKFA1H-rDdc7VOtvEVpUJ82TwyQ/rs:fill:1000:1000:1:0/plain/bafkreiaivizp4xldojmmpuzmiu75cmea7nq56dnntnuhzhsjcb63aou5ei@jpeg",
          "declaration": {
            "actorType": "app.bsky.system.actorUser",
            "cid": "cids(1)",
          },
          "did": "user(0)",
          "displayName": "ali",
          "handle": "alice.test",
          "viewer": {
            "muted": false,
          },
        },
        "cid": "cids(0)",
        "downvoteCount": 1,
        "indexedAt": "1970-01-01T00:00:00.000Z",
        "record": {
          "$type": "app.bsky.feed.post",
          "createdAt": "1970-01-01T00:00:00.000Z",
          "text": "again",
        },
        "replyCount": 2,
        "repostCount": 1,
        "upvoteCount": 2,
        "uri": "record(0)",
        "viewer": {},
      },
    },
  },
  {
    "post": {
      "author": {
        "avatar": "https://pds.public.url/image/KzkHFsMRQ6oAKCHCRKFA1H-rDdc7VOtvEVpUJ82TwyQ/rs:fill:1000:1000:1:0/plain/bafkreiaivizp4xldojmmpuzmiu75cmea7nq56dnntnuhzhsjcb63aou5ei@jpeg",
        "declaration": {
          "actorType": "app.bsky.system.actorUser",
          "cid": "cids(1)",
        },
        "did": "user(0)",
        "displayName": "ali",
        "handle": "alice.test",
        "viewer": {
          "muted": false,
        },
      },
      "cid": "cids(7)",
      "downvoteCount": 1,
      "embed": {
        "$type": "app.bsky.embed.record#presented",
        "record": {
          "$type": "app.bsky.embed.record#presentedRecord",
          "author": {
            "declaration": {
              "actorType": "app.bsky.system.actorUser",
              "cid": "cids(1)",
            },
            "did": "user(1)",
            "handle": "dan.test",
            "viewer": {
              "following": "record(1)",
              "muted": false,
            },
          },
          "cid": "cids(2)",
          "record": {
            "$type": "app.bsky.feed.post",
            "createdAt": "1970-01-01T00:00:00.000Z",
            "entities": [
              {
                "index": {
                  "end": 18,
                  "start": 0,
                },
                "type": "mention",
                "value": "user(0)",
              },
            ],
            "text": "@alice.bluesky.xyz is the best",
          },
          "uri": "record(2)",
        },
      },
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "record": {
        "$type": "app.bsky.feed.post",
        "createdAt": "1970-01-01T00:00:00.000Z",
        "embed": {
          "$type": "app.bsky.embed.record",
          "record": {
            "cid": "cids(2)",
            "uri": "record(2)",
          },
        },
        "text": "yoohoo",
      },
      "replyCount": 0,
      "repostCount": 0,
      "upvoteCount": 1,
      "uri": "record(10)",
      "viewer": {},
    },
  },
  {
    "post": {
      "author": {
        "avatar": "https://pds.public.url/image/KzkHFsMRQ6oAKCHCRKFA1H-rDdc7VOtvEVpUJ82TwyQ/rs:fill:1000:1000:1:0/plain/bafkreiaivizp4xldojmmpuzmiu75cmea7nq56dnntnuhzhsjcb63aou5ei@jpeg",
        "declaration": {
          "actorType": "app.bsky.system.actorUser",
          "cid": "cids(1)",
        },
        "did": "user(3)",
        "displayName": "bobby",
        "handle": "bob.test",
        "viewer": {
          "followedBy": "record(8)",
          "following": "record(7)",
          "muted": false,
        },
      },
      "cid": "cids(8)",
      "downvoteCount": 0,
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "record": {
        "$type": "app.bsky.feed.post",
        "createdAt": "1970-01-01T00:00:00.000Z",
        "text": "bobby boy here",
      },
      "replyCount": 0,
      "repostCount": 0,
      "upvoteCount": 0,
      "uri": "record(11)",
      "viewer": {},
    },
  },
  {
    "post": {
      "author": {
        "avatar": "https://pds.public.url/image/KzkHFsMRQ6oAKCHCRKFA1H-rDdc7VOtvEVpUJ82TwyQ/rs:fill:1000:1000:1:0/plain/bafkreiaivizp4xldojmmpuzmiu75cmea7nq56dnntnuhzhsjcb63aou5ei@jpeg",
        "declaration": {
          "actorType": "app.bsky.system.actorUser",
          "cid": "cids(1)",
        },
        "did": "user(0)",
        "displayName": "ali",
        "handle": "alice.test",
        "viewer": {
          "muted": false,
        },
      },
      "cid": "cids(0)",
      "downvoteCount": 1,
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "record": {
        "$type": "app.bsky.feed.post",
        "createdAt": "1970-01-01T00:00:00.000Z",
        "text": "again",
      },
      "replyCount": 2,
      "repostCount": 1,
      "upvoteCount": 2,
      "uri": "record(0)",
      "viewer": {},
    },
  },
  {
    "post": {
      "author": {
        "declaration": {
          "actorType": "app.bsky.system.actorUser",
          "cid": "cids(1)",
        },
        "did": "user(1)",
        "handle": "dan.test",
        "viewer": {
          "following": "record(1)",
          "muted": false,
        },
      },
      "cid": "cids(9)",
      "downvoteCount": 0,
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "record": {
        "$type": "app.bsky.feed.post",
        "createdAt": "1970-01-01T00:00:00.000Z",
        "text": "dan here!",
      },
      "replyCount": 0,
      "repostCount": 0,
      "upvoteCount": 0,
      "uri": "record(12)",
      "viewer": {},
    },
  },
  {
    "post": {
      "author": {
        "declaration": {
          "actorType": "app.bsky.system.actorUser",
          "cid": "cids(1)",
        },
        "did": "user(2)",
        "handle": "carol.test",
        "viewer": {
          "followedBy": "record(4)",
          "following": "record(3)",
          "muted": false,
        },
      },
      "cid": "cids(10)",
      "downvoteCount": 0,
      "embed": {
        "$type": "app.bsky.embed.images#presented",
        "images": [
          {
            "alt": "tests/image/fixtures/key-landscape-small.jpg",
            "fullsize": "https://pds.public.url/image/AiDXkxVbgBksxb1nfiRn1m6S4K8_mee6o8r-UGLNzOM/rs:fit:2000:2000:1:0/plain/bafkreigy5p3xxceipk2o6nqtnugpft26ol6yleqhboqziino7axvdngtci@jpeg",
            "thumb": "https://pds.public.url/image/uc7FGfiGv0mMqmk9XiqHXrIhNymLHaex7Ge8nEhmXqo/rs:fit:1000:1000:1:0/plain/bafkreigy5p3xxceipk2o6nqtnugpft26ol6yleqhboqziino7axvdngtci@jpeg",
          },
          {
            "alt": "tests/image/fixtures/key-alt.jpg",
            "fullsize": "https://pds.public.url/image/xC2No-8rKVDIwIMmCiEBm9EiGLDBBOpf36PHoGf-GDw/rs:fit:2000:2000:1:0/plain/bafkreifdklbbcdsyanjz3oqe5pf2omuq5ansthokxlbleagg3eenx62h7e@jpeg",
            "thumb": "https://pds.public.url/image/g7yazUpNwN8LKumZ2Zmn_ptQbtMLs1Pti5-GDn7H8_8/rs:fit:1000:1000:1:0/plain/bafkreifdklbbcdsyanjz3oqe5pf2omuq5ansthokxlbleagg3eenx62h7e@jpeg",
          },
        ],
      },
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "record": {
        "$type": "app.bsky.feed.post",
        "createdAt": "1970-01-01T00:00:00.000Z",
        "embed": {
          "$type": "app.bsky.embed.images",
          "images": [
            {
              "alt": "tests/image/fixtures/key-landscape-small.jpg",
              "image": {
                "cid": "cids(5)",
                "mimeType": "image/jpeg",
              },
            },
            {
              "alt": "tests/image/fixtures/key-alt.jpg",
              "image": {
                "cid": "cids(11)",
                "mimeType": "image/jpeg",
              },
            },
          ],
        },
        "text": "hi im carol",
      },
      "replyCount": 0,
      "repostCount": 0,
      "upvoteCount": 2,
      "uri": "record(13)",
      "viewer": {
        "upvote": "record(14)",
      },
    },
  },
  {
    "post": {
      "author": {
        "avatar": "https://pds.public.url/image/KzkHFsMRQ6oAKCHCRKFA1H-rDdc7VOtvEVpUJ82TwyQ/rs:fill:1000:1000:1:0/plain/bafkreiaivizp4xldojmmpuzmiu75cmea7nq56dnntnuhzhsjcb63aou5ei@jpeg",
        "declaration": {
          "actorType": "app.bsky.system.actorUser",
          "cid": "cids(1)",
        },
        "did": "user(3)",
        "displayName": "bobby",
        "handle": "bob.test",
        "viewer": {
          "followedBy": "record(8)",
          "following": "record(7)",
          "muted": false,
        },
      },
      "cid": "cids(12)",
      "downvoteCount": 0,
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "record": {
        "$type": "app.bsky.feed.post",
        "createdAt": "1970-01-01T00:00:00.000Z",
        "text": "bob back at it again!",
      },
      "replyCount": 0,
      "repostCount": 0,
      "upvoteCount": 0,
      "uri": "record(15)",
      "viewer": {},
    },
  },
  {
    "post": {
      "author": {
        "avatar": "https://pds.public.url/image/KzkHFsMRQ6oAKCHCRKFA1H-rDdc7VOtvEVpUJ82TwyQ/rs:fill:1000:1000:1:0/plain/bafkreiaivizp4xldojmmpuzmiu75cmea7nq56dnntnuhzhsjcb63aou5ei@jpeg",
        "declaration": {
          "actorType": "app.bsky.system.actorUser",
          "cid": "cids(1)",
        },
        "did": "user(0)",
        "displayName": "ali",
        "handle": "alice.test",
        "viewer": {
          "muted": false,
        },
      },
      "cid": "cids(13)",
      "downvoteCount": 0,
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "record": {
        "$type": "app.bsky.feed.post",
        "createdAt": "1970-01-01T00:00:00.000Z",
        "text": "hey there",
      },
      "replyCount": 0,
      "repostCount": 0,
      "upvoteCount": 0,
      "uri": "record(16)",
      "viewer": {},
    },
  },
]
`;
